examples/with-react-intl/pages/_app.tsx

```diff
@@ -5,37 +5,73 @@ import App from 'next/app';
 
 function MyApp({Component, pageProps, locale, messages}) {
   return (
-    <IntlProvider locale={locale} messages={messages}>
+    <IntlProvider locale={locale} defaultLocale="en" messages={messages}>
       <Component {...pageProps} />
     </IntlProvider>
   );
 }
 
-// We need to load and expose the translations on the request for the user's
-// locale. These will only be used in production, in dev the `defaultMessage` in
-// each message description in the source code will be used.
-const getMessages = (locale: string = 'en') => {
-  switch (locale) {
-    default:
-      return import('../compiled-lang/en.json');
-    case 'fr':
-      return import('../compiled-lang/fr.json');
+/**
+ * Get the messages and also do locale negotiation. A multi-lingual user
+ * can specify locale prefs like ['ja', 'en-GB', 'en'] which is interpreted as
+ * Japanese, then British English, then English
+ * @param locales list of requested locales
+ * @returns {[string, Promise]} A tuple containing the negotiated locale
+ * and the promise of fetching the translated messages
+ */
+function getMessages(locales: string | string[] = ['en']) {
+  if (!Array.isArray(locales)) {
+    locales = [locales];
   }
-};
+  let langBundle;
+  let locale;
+  for (let i = 0; i < locales.length && !locale; i++) {
+    locale = locales[i];
+    switch (locale) {
+      case 'fr':
+        langBundle = import('../compiled-lang/fr.json');
```

Comment:
Is there a reason to hard-code the imports here, instead of something like:
```js
langBundle = import(`../compiled-lang/${locale}.json`)
```
and have webpack take care of this?

Reply:
What happens if you don't support that locale?

Reply:
If I remember correctly, the Promise to load the specific module would reject. Which should be similar to the current fallback solution. Maybe with a check to see if the requested locale is part of the list of supported locales to short-circuit the detection.

So far, I did not encounter this issue at runtime, as I'm usually having an enum-type object upstream that limits the selection so only known values are inputs of the function. In this case I would add an `supportedLocales.includes(locale)` type of check to make sure, only supported locales are loaded.
That check could be powered by `require.context` to get the list of supported locales directly from the available locales.

Reply:
Yeah ￼￼￼￼my concern is maintaining supportedLocales separately. Does require.context work in client side though?

Reply:
`require.context` will be resolved at build time. It works client-side via the webpack runtime but it doesn't recognize new files during runtime, as it is based on the layout of the file tree at build.

Reply:
On the client it is basically a specially prepared version of `require`.
I have successfully used it to load components dynamically based on runtime config like CMS-content. So I'm confident this works, at least for this case here.
As a small bonus, we can use the same code to load messages etc. during development

Reply:
maybe @lfades can advise here from a next.js perspective. IMO relying on `require` is basically relying on `webpack` implementation details instead of spec'ed runtime environment (such as ECMA262 or such). Different runtimes can deal with `require` differently (e.g `jest` does module resolution in a custom manner...) so it's fairly unstable. TBH idk what `next`'s guarantee of this behavior is since it is undocumented.

Reply:
`import` is indeed much more preferred here for the reasons that @longlho mentioned